File: GUI.java
// Copyright (C) 2021 Jarmo Hurri

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

import javax.swing.*;
import java.awt.FlowLayout;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.time.LocalTime;

// gui extends frame and inherits all its methods (e.g., for setting
// size, adding components)

public class GUI extends JFrame
{
  // constructor will create all GUI components and set their properties
  public GUI (Application application)
  {
    super ("A simple GUI demo");
    this.application = application;

    // frame properties
    setSize (400, 300);
    setLayout (new FlowLayout ());

    // menu bar with one item for quitting the program
    menuBar = new JMenuBar ();
    JMenu fileMenu = new JMenu ("File");
    JMenuItem quitItem = new JMenuItem ("Quit");
    quitItem.addActionListener ((ActionEvent e) ->
                                System.exit (0));
    fileMenu.add (quitItem);
    menuBar.add (fileMenu);
    setJMenuBar (menuBar);

    // a label to show time as string, and timer for updating string
    timeLabel = new JLabel ();
    add (timeLabel);

    // a label for showing the boolean state of the program
    stateLabel = new JLabel ();
    showState ();
    add (stateLabel);

    // a button for flipping program state
    flipButton = new JButton ("Flip");
    flipButton.addActionListener ((ActionEvent e) ->
                                  {
                                    application.flipState ();
                                    showState ();
                                  });
    add (flipButton);
    // progress bar tracking elapsed time
    progressBar = new JProgressBar (0, TMAX);
    add (progressBar);
    showTime ();

    setVisible (true); // show frame
  }

  // a method which shows the state 
  private void showState ()
  {
    stateLabel.setText (String.valueOf (application.getState ()));
  }

  public void updateAndShowTime ()
  {
    secs++;
    secs %= TMAX;
    showTime ();
  }
  
  private void showTime ()
  {
    LocalTime now = LocalTime.now ();
    int hours = now.getHour (), minutes = now.getMinute (), seconds = now.getSecond ();
    timeLabel.setText (formatInt (hours) + ":" + formatInt (minutes) + ":" + formatInt (seconds));
    progressBar.setValue (secs);
  }

  // format an int as a string so that the minimum length is 2
  // characters, zero-padded if needed
  private String formatInt (int i)
  {
    return String.format ("%02d", i);
  }
  
  private Application application;
  private JMenuBar menuBar;
  private JLabel stateLabel, timeLabel;
  private JButton flipButton;
  private JProgressBar progressBar;
  private final int TMAX = 30;
  private int secs = 0;
}

File: GUI.py
from PySide6.QtWidgets import QMainWindow, QHBoxLayout, QWidget, QLabel, QPushButton, QProgressBar
from PySide6.QtGui import QAction
import time

class GUI(QMainWindow):
    def __init__(self, application):
        super().__init__()
        self.application = application

        # menubar with one menu, one action to quit
        file_menu = self.menuBar().addMenu('File')
        quit_action = QAction("Quit", self)
        quit_action.triggered.connect(application.quit)
        file_menu.addAction(quit_action)

        # horizontal layout for components
        layout = QHBoxLayout()
        layout_widget = QWidget()
        layout_widget.setLayout(layout)
        self.setCentralWidget(layout_widget)
        
        # label showing time
        self.time_label = QLabel(layout_widget)
        layout.addWidget(self.time_label)

        # label showing state
        self.state_label = QLabel(layout_widget)
        layout.addWidget(self.state_label)

        # button for flipping state
        button = QPushButton('Flip', layout_widget)
        layout.addWidget(button)
        button.clicked.connect(self._flip_state)

        # cycling progress bar
        self.COUNTER_MAX = 30
        self.counter = 0
        self.progress_bar = QProgressBar(layout_widget)
        self.progress_bar.setMinimum(0)
        self.progress_bar.setMaximum(self.COUNTER_MAX - 1)
        layout.addWidget(self.progress_bar)

        self._show_state()
        self._show_time()
        self.show()

    # update counter and show time and counter status
    def update_and_show_time(self):
        self.counter += 1
        self.counter %= self.COUNTER_MAX
        self._show_time()

    # show the state of the app
    def _show_state(self):
        self.state_label.setText(f'{self.application.get_state()}')

    # flip the state of the app
    def _flip_state(self):
        self.application.flip_state()
        self._show_state()

    # show time and counter status
    def _show_time(self):
        self.time_label.setText(f'{time.strftime("%H:%M:%S")}')
        self.progress_bar.setValue(self.counter)
